filters/openpolicyagent/openpolicyagent.go

```diff
@@ -476,38 +477,74 @@ func (registry *OpenPolicyAgentRegistry) new(store storage.Store, configBytes []
 // Start asynchronously starts the policy engine's plugins that download
 // policies, report status, etc.
 func (opa *OpenPolicyAgentInstance) Start(ctx context.Context, timeout time.Duration) error {
-	err := opa.manager.Start(ctx)
+	discoveryPlugin := discovery.Lookup(opa.manager)
 
-	if err != nil {
-		return err
-	}
+	done := make(chan struct{})
+	failed := make(chan error, 1)
 
-	// check readiness of all plugins
-	pluginsReady := func() bool {
-		for _, status := range opa.manager.PluginStatus() {
-			if status != nil && status.State != plugins.StateOK {
-				return false
+	discoveryPlugin.RegisterListener("startuplistener", func(status bundle.Status) {
+		handleStatusErrors(status, failed, "discovery plugin")
+	})
+
+	var registerBundleListenerOnce sync.Once
+
+	// Register listener for "bundle" plugin
+	opa.manager.RegisterPluginStatusListener("bundlelistener", func(status map[string]*plugins.Status) {
+		if _, exists := status["bundle"]; exists {
+			bundlePlugin := bundle.Lookup(opa.manager)
+			if bundlePlugin != nil {
+				registerBundleListenerOnce.Do(func() {
+					bundlePlugin.Register("startuplistener", func(status bundle.Status) {
+						handleStatusErrors(status, failed, "bundle plugin")
+					})
+
+				})
 			}
 		}
-		return true
-	}
+	})
+	defer opa.manager.UnregisterPluginStatusListener("bundlelistener")
 
-	err = waitFunc(ctx, pluginsReady, 100*time.Millisecond)
+	// Register listener for general plugin status checks
+	opa.manager.RegisterPluginStatusListener("generalStartUpListener", func(status map[string]*plugins.Status) {
+		for _, pluginStatus := range status {
+			if pluginStatus != nil && pluginStatus.State != plugins.StateOK {
+				return
+			}
+		}
+		done <- struct{}{}
+	})
+	defer opa.manager.UnregisterPluginStatusListener("generalStartUpListener")
 
+	err := opa.manager.Start(ctx)
 	if err != nil {
+		return err
+	}
+
+	select {
+	case <-done:
+		return nil
+	case err := <-failed:
+		opa.Close(ctx)
+
+		return err
+	case <-ctx.Done():
+		timeoutErr := ctx.Err()
```

Comment:
Here in context canceled or times-out case, we should clean up resources (opa.Close(ctx)) and return the appropriate error.

Reply:
As far as I see that will be done in line 542.